src/lib.rs

```diff
@@ -9,11 +9,151 @@
+    }
+    fn no_input(&mut self,
+                matches: &getopts::Matches,
+                sopts: &config::Options,
+                odir: &Option<PathBuf>,
+                ofile: &Option<PathBuf>,
+                descriptions: &diagnostics::registry::Registry)
+                -> Option<(Input, Option<PathBuf>)> {
+        self.0.no_input(matches, sopts, odir, ofile, descriptions)
+    }
+    fn late_callback(&mut self,
+                     matches: &getopts::Matches,
+                     sess: &Session,
+                     input: &Input,
+                     odir: &Option<PathBuf>,
+                     ofile: &Option<PathBuf>)
+                     -> Compilation {
+        self.0.late_callback(matches, sess, input, odir, ofile)
+    }
+    fn build_controller(&mut self, sess: &Session) -> driver::CompileController<'a> {
+        let mut control = self.0.build_controller(sess);
+
+        let old = std::mem::replace(&mut control.after_parse.callback, box |_| {});
+        control.after_parse.callback = Box::new(move |state| {
+            {
+                let mut registry = rustc_plugin::registry::Registry::new(state.session, state.krate.expect("at this compilation stage the krate must be parsed"));
+                registry.args_hidden = Some(Vec::new());
+                plugin_registrar(&mut registry);
+
+                let rustc_plugin::registry::Registry { early_lint_passes, late_lint_passes, lint_groups, llvm_passes, attributes, mir_passes, .. } = registry;
+                let sess = &state.session;
+                let mut ls = sess.lint_store.borrow_mut();
+                for pass in early_lint_passes {
+                    ls.register_early_pass(Some(sess), true, pass);
+                }
+                for pass in late_lint_passes {
+                    ls.register_late_pass(Some(sess), true, pass);
+                }
+
+                for (name, to) in lint_groups {
+                    ls.register_group(Some(sess), true, name, to);
+                }
+
+                sess.plugin_llvm_passes.borrow_mut().extend(llvm_passes);
+                sess.mir_passes.borrow_mut().extend(mir_passes);
+                sess.plugin_attributes.borrow_mut().extend(attributes);
+            }
+            old(state);
+        });
+
+        control
+    }
+}
+
+use std::path::Path;
+
+pub fn main() {
+    use std::env;
+
+    if env::var("DOGFOOD").map(|_| true).unwrap_or(false) {
+        return;
+    }
+
+    let dep_path = env::current_dir().expect("current dir is not readable").join("target").join("debug").join("deps");
+    let sys_root = match (option_env!("MULTIRUST_HOME"), option_env!("MULTIRUST_TOOLCHAIN")) {
+        (Some(home), Some(toolchain)) => format!("{}/toolchains/{}", home, toolchain),
+        _ => option_env!("SYSROOT").expect("need to specify SYSROOT env var during clippy compilation").to_owned(),
+    };
+
+    if env::var("AUTOBOTS").map(|s| s == "ROLLOUT").unwrap_or(false) {
```

Comment:
this is the magic behind it all. The `cargo-clippy' executable that is installed forwards to `cargo rustc` with the `RUSTC` env flag set to the executable itself. We'd get into an infinite loop now if we'd not also set some special flag that ensures we know we should run clippy instead of `cargo rustc`. So I added an arbitrary flag. We could do other things like check for the `RUSTC` env flag, but that might interact badly with other code.

name can be arbitrarily bikeshedded, I just needed something that definitely won't collide with other env vars